kafka-rest/src/main/java/io/confluent/kafkarest/controllers/ControllersModule.java

```diff
@@ -38,9 +49,72 @@ protected void configure() {
     bind(ReassignmentManagerImpl.class).to(ReassignmentManager.class);
     bind(RecordSerializerFacade.class).to(RecordSerializer.class);
     bind(ReplicaManagerImpl.class).to(ReplicaManager.class);
-    bind(SchemaManagerImpl.class).to(SchemaManager.class);
-    bindAsContract(SchemaRecordSerializer.class);
+    bindFactory(SchemaManagerFactory.class).to(SchemaManager.class);
     bind(TopicConfigManagerImpl.class).to(TopicConfigManager.class);
     bind(TopicManagerImpl.class).to(TopicManager.class);
+    bindFactory(SchemaRecordSerializerFactory.class).to(SchemaRecordSerializer.class);
+  }
+
+  private static final class SchemaRecordSerializerFactory
+      implements Factory<SchemaRecordSerializer> {
+
+    private final Optional<SchemaRegistryClient> schemaRegistryClient;
+    private final Map<String, Object> avroSerializerConfigs;
+    private final Map<String, Object> jsonschemaSerializerConfigs;
+    private final Map<String, Object> protobufSerializerConfigs;
+
+    @Inject
+    private SchemaRecordSerializerFactory(
+        Optional<SchemaRegistryClient> schemaRegistryClient,
+        @AvroSerializerConfigs Map<String, Object> avroSerializerConfigs,
+        @JsonschemaSerializerConfigs Map<String, Object> jsonschemaSerializerConfigs,
+        @ProtobufSerializerConfigs Map<String, Object> protobufSerializerConfigs) {
+      this.schemaRegistryClient = requireNonNull(schemaRegistryClient);
+      this.avroSerializerConfigs = requireNonNull(avroSerializerConfigs);
+      this.jsonschemaSerializerConfigs = requireNonNull(jsonschemaSerializerConfigs);
+      this.protobufSerializerConfigs = requireNonNull(protobufSerializerConfigs);
+    }
+
+    @Override
+    public SchemaRecordSerializer provide() {
+      if (schemaRegistryClient.isPresent()) {
+        return new SchemaRecordSerializerImpl(
+            schemaRegistryClient.get(),
+            avroSerializerConfigs,
+            jsonschemaSerializerConfigs,
+            protobufSerializerConfigs);
+      } else {
+        return new SchemaRecordSerializerThrowing();
+      }
+    }
+
+    @Override
+    public void dispose(SchemaRecordSerializer schemaRecordSerializer) {}
+  }
+
+  private static final class SchemaManagerFactory implements Factory<SchemaManager> {
+
+    private final Optional<SchemaRegistryClient> schemaRegistryClient;
+    private final SubjectNameStrategy defaultSubjectNameStrategy;
+
+    @Inject
+    private SchemaManagerFactory(
+        Optional<SchemaRegistryClient> schemaRegistryClient,
+        SubjectNameStrategy defaultSubjectNameStrategy) {
+      this.schemaRegistryClient = schemaRegistryClient;
```

Comment:
requireNonNull

Reply:
That's odd, I have that locally.  Will do a dummy commit and repush